text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+        self.into_iter()
+    }
+}
+
+// becomes
+
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    
+    type $ = impl Iterator<Item = Self::Item>;
+
+    fn into_iter(self) -> <Self as NewIntoIterator>::$ {
+        self.into_iter()
+    }
+}
+```
+
+
+## Generic parameter capture and GATs
+
+Given a trait method with a return type like `-> impl A + ... + Z` and an implementation of that trait, the hidden type for that implementation is allowed to reference:
+
+* Concrete types, constant expressions, and `'static`
+* `Self`
+* Generics on the impl
+* Certain generics on the method
+    * Explicit type parameters
+    * Argument-position `impl Trait` types
+    * Explicit const parameters
+    * Lifetime parameters that appear anywhere in `A + ... + Z`, including elided lifetimes
+
+We say that a generic parameter is *captured* if it may appear in the hidden type. These rules are the same as those for `-> impl Trait` in inherent impls.
+
+When desugaring, captured parameters from the method are reflected as generic parameters on the `$` associated type. Furthermore, the `$` associated type brings whatever where clauses are declared on the method into scope, excepting those which reference parameters that are not captured.
+
+This transformation is precisely the same as the one which is applied to other forms of `-> impl Trait`, except that it applies to an associated type and not a top-level type alias. 
+
+Example:
+
+```rust
+trait RefIterator for Vec<u32> {
+    type Item<'me>
+    where 
+        Self: 'me;
+
+    fn iter<'a>(&'a self) -> impl Iterator<Item = Self:Item<'a>>;
+}
+
+// Since 'a is named in the bounds, it is captured.
+// `RefIterator` thus becomes:
+
+trait RefIterator for Vec<u32> {
+    type Item<'me>
+    where 
+        Self: 'me;
+
+    type $<'a>: impl Iterator<Item = Self::Item<'a>>
+    where 
+        Self: 'a; // Implied bound from fn
+
+    fn iter<'a>(&'a self) -> Self::$<'a>;
+}
+```
+
+## Validity constraint on impls
+
+Given a trait method where `impl Trait` appears in return position,
+
+```rust
+trait Trait {
+    fn method() -> impl T_0 + ... + T_m;
+}
+```
+
+where `T_0 + ... + T_m` are bounds, for any impl of that trait to be valid, the following conditions must hold:
+
+* The return type named in the corresponding impl method must implement all bounds `T_0 + ... + T_m` specified in the trait.
+* Either the impl method must have `#[refine]`,[^refine] OR
+    * The impl must use `impl Trait` syntax to name the corresponding type, and
+    * The return type in the trait must implement all bounds `I_0 + ... + I_n` specified in the impl return type. (Taken with the first outer bullet point, we can say that the bounds in the trait and the bounds in the impl imply each other.)
+
+[^refine]: `#[refine]` was added in [RFC 3245: Refined trait implementations](https://rust-lang.github.io/rfcs/3245-refined-impls.html). This feature is not yet stable.
```

Comment:
Does this mean this RFC is dependent on 3245 being stabilized first, or is the plan to include a subset of 3245 in this RFC?

Reply:
Added clarifying text to this footnote; it just means any code that requires `#[refine]` won't work until that feature is stabilized. Simple examples without refinement will continue to work.